### PR TITLE
improved subset type errors, export set name space, cloning bug fix

### PR DESCRIPTION
**Pull request by RustanLeino**
I had these 3 changes in the same branch, so they're apparently going into the same pull request. (Next time, I'll separate unrelated pull requests.)

### CODE REVIEW

File: Source/Dafny/Cloner.cs [hunk 1]
```diff
@@ -601,9 +601,6 @@ class Cloner
       } else if (stmt is ForallStmt) {
         var s = (ForallStmt)stmt;
         r = new ForallStmt(Tok(s.Tok), Tok(s.EndTok), s.BoundVars.ConvertAll(CloneBoundVar), null, CloneExpr(s.Range), s.Ens.ConvertAll(CloneMayBeFreeExpr), CloneStmt(s.Body));
-        if (s.ForallExpressions != null) {
```
Comment (seanmcl) [inline Source/Dafny/Cloner.cs]: When do expressions need to be cloned?

File: Source/Dafny/DafnyAst.cs [hunk 2]
```diff
@@ -3416,14 +3417,7 @@ public class ModuleSignature {
...
     public bool FindExport(string name, out ModuleExportDecl pp) {
-      TopLevelDecl top;
-      if (TopLevels.TryGetValue(name, out top) && top is ModuleExportDecl) {
-        pp = ((ModuleExportDecl)top);
-        return true;
-      } else {
-        pp = null;
-        return false;
-      }
+      return ExportSets.TryGetValue(name, out pp);
```
Comment (seanmcl) [inline Source/Dafny/DafnyAst.cs]: Can we just delete this method?  If you're exporting `ExportSets` as `public` anyway, this seems unnecessary.

File: Source/Dafny/Cloner.cs (hunk 1 above)
Comment (RustanLeino) [reply Source/Dafny/Cloner.cs]: Whenever there's a need to create a new syntactic copy of a part of an AST. Examples are:
* when a module is refined, its contents is cloned and included into the new module
* when the program is compiled, all of its modules get cloned before they are compiled (this used to be needed, but I'm not actually sure there's any reason to still do this)
* when the various export sets of a module are checked for consistency, the exported components are cloned
* when an `inductive predicate`, `copredicate`, `inductive lemma`, or `colemma` is syntactically desugared, parts of it are cloned

File: Source/Dafny/DafnyAst.cs (hunk 2 above)
Comment (RustanLeino) [reply Source/Dafny/DafnyAst.cs]: Good idea.

**Pull request merged**